src/valid/analyzer.rs

```diff
@@ -159,6 +159,24 @@ impl ExpressionInfo {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize))]
+#[cfg_attr(feature = "deserialize", derive(serde::Deserialize))]
+enum GlobalOrArgument {
+    Global(Handle<crate::GlobalVariable>),
+    Argument(u32),
+}
+
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize))]
+#[cfg_attr(feature = "deserialize", derive(serde::Deserialize))]
+struct ArgumentSampling {
+    /// Whether this argument is used as an image or a sampler
```

Comment:
why do we need to track this? Since we know the type of the argument, it's always unambiguous if it's used as a sampler or as an image